pkg/ipfs/repository.go

```diff
@@ -0,0 +1,142 @@
+// VulcanizeDB
+// Copyright © 2019 Vulcanize
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package ipfs
+
+import (
+	"github.com/jmoiron/sqlx"
+	"github.com/lib/pq"
+
+	"github.com/vulcanize/vulcanizedb/pkg/datastore/postgres"
+)
+
+// CIDRepository is an interface for indexing CIDPayloads
+type CIDRepository interface {
+	Index(cidPayload *CIDPayload) error
+}
+
+// Repository is the underlying struct for the CIDRepository interface
+type Repository struct {
+	db *postgres.DB
+}
+
+// NewCIDRepository creates a new pointer to a Repository which satisfies the CIDRepository interface
+func NewCIDRepository(db *postgres.DB) *Repository {
+	return &Repository{
+		db: db,
+	}
+}
+
+// Index indexes a cidPayload in Postgres
+func (repo *Repository) Index(cidPayload *CIDPayload) error {
+	tx, err := repo.db.Beginx()
+	if err != nil {
+		return err
+	}
+	headerID, err := repo.indexHeaderCID(tx, cidPayload.HeaderCID, cidPayload.BlockNumber, cidPayload.BlockHash.Hex())
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	for uncleHash, cid := range cidPayload.UncleCIDS {
+		err = repo.indexUncleCID(tx, cid, cidPayload.BlockNumber, uncleHash.Hex())
+		if err != nil {
+			tx.Rollback()
+			return err
+		}
+	}
+	err = repo.indexTransactionAndReceiptCIDs(tx, cidPayload, headerID)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	err = repo.indexStateAndStorageCIDs(tx, cidPayload, headerID)
+	if err != nil {
+		tx.Rollback()
```

Comment:
I know it can be a little tedious, but do you think we want to be assigning possible errors on `tx.Rollback()` so that we can include them in the logs?

Reply:
That seems like a good idea. I have poor support for error logging throughout most this code, I'll work on this.

Reply:
Sorry, to clarify here, do you mean to capture the potential errors of the rollback itself
```go
rollBackErr := tx.Rollback()
```
?

Reply:
@i-norden Yes. I think we just log them today, nothing we can do